src/coffea/dataset_tools/preprocess.py

```diff
@@ -0,0 +1,169 @@
+import numpy
+import uproot
+
+
+def _get_steps(
+    normed_files,
+    maybe_step_size=None,
+    align_clusters=False,
+    recalculate_seen_steps=False,
+    skip_bad_files=False,
+    file_exceptions=(FileNotFoundError, OSError),
+):
+    nf_backend = awkward.backend(normed_files)
+    lz_or_nf = awkward.typetracer.length_zero_if_typetracer(normed_files)
+
+    array = [] if nf_backend != "typetracer" else lz_or_nf
+    for arg in lz_or_nf:
+        try:
+            the_file = uproot.open({arg.file: None})
+        except file_exceptions as e:
+            if skip_bad_files:
+                array.append(None)
+                continue
+            else:
+                raise e
+
+        tree = the_file[arg.object_path]
+        num_entries = tree.num_entries
+
+        target_step_size = num_entries if maybe_step_size is None else maybe_step_size
+
+        file_uuid = str(the_file.file.uuid)
+
+        out_uuid = arg.uuid
+        out_steps = arg.steps
+
+        if out_uuid != file_uuid or recalculate_seen_steps:
+            if align_clusters:
+                clusters = tree.common_entry_offsets()
+                out = [0]
+                for c in clusters:
+                    if c >= out[-1] + target_step_size:
+                        out.append(c)
+                if clusters[-1] != out[-1]:
+                    out.append(clusters[-1])
+                out = numpy.array(out, dtype="int64")
+                out = numpy.stack((out[:-1], out[1:]), axis=1)
+            else:
+                n_steps = math.ceil(num_entries / target_step_size)
+                out = numpy.array(
+                    [
+                        [
+                            i * target_step_size,
+                            min((i + 1) * target_step_size, num_entries),
+                        ]
+                        for i in range(n_steps)
+                    ],
+                    dtype="int64",
+                )
+
+            out_uuid = file_uuid
+            out_steps = out.tolist()
+
+        array.append(
+            {
+                "file": arg.file,
+                "object_path": arg.object_path,
+                "steps": out_steps,
+                "uuid": out_uuid,
+            }
+        )
+
+    if len(array) == 0:
+        array = awkward.Array(
+            [
+                {"file": "junk", "object_path": "junk", "steps": [[]], "uuid": "junk"},
+                None,
+            ]
+        )
+        array = awkward.Array(array.layout.form.length_zero_array(highlevel=False))
+    else:
+        array = awkward.Array(array)
+
+    if nf_backend == "typetracer":
+        array = awkward.Array(
+            array.layout.to_typetracer(forget_length=True),
+        )
+
+    return array
+
+
+def preprocess(
+    fileset,
+    maybe_step_size=None,
+    align_clusters=False,
+    recalculate_seen_steps=False,
+    files_per_batch=1,
+    skip_bad_files=False,
+    file_exceptions=(FileNotFoundError, OSError),
+):
```

Comment:
Could the return type be described by a dataclass or protocol? (do we want to return a proper class here?)
We could normalize all instances to a proper dataclass on input to the various functions.

Reply:
The return type here is actually the same as the input type but in the case of the "runnable" output all Optional types are removed.

Reply:
@nsmith- I've added typing here now to define the interface more concretely - please have a look.